### ISSUE TITLE
Blake hashing support

### ISSUE THREAD

**Issue by Ericson2314** [NixOS/nix#11999]
The Blake family of hash algorithms has a number of advantages for random access because data is ingested in a tree rather than linear stream. Tvix is using it for this purpose.

In order to support such experiments and potential compatibility with Tvix's experiments in the future, I think we should the same Blake hashing algorithm that they do.

CC @edef1c

**Comment by silvanshade** [comment]
I have some thoughts on this based on experience implementing support for BLAKE3 in my personal Nix fork.

I submitted an initial version based on the single-threaded C BLAKE3 implementation in the PR https://github.com/NixOS/nix/pull/10600

I opted to close the PR shortly after since I had another version that used the Rust-based multi-threaded Rayon implementation which was much faster under the expected circumstances.

Ultimately, I never submitted a second PR using the Rust version for a few reasons.

The Rust version required a fairly substantial amount of additional infrastructure code in order to be able to both define bidirectional C++ <-> Rust bindings (in a clean, extensible way) and also to be able to support building and linking the Rust code along with the rest of the C++ build process.

There was also some refactoring I had to do with how the hashing API was defined on the C++ side since, in order to take advantage of the full performance of the Rust implementation, I needed to allow the Rust implementation to handle IO. The default behavior of the Nix implementation was to use chunks that were too small to take advantage of parallelism and also did not provide support for memory-mapped IO which is the fastest scenario for the Rust BLAKE3 implementation.

I solved these issues and had a working version but at the time I was unsure about whether it was a good idea to submit something potentially that substantial just in order to have a more performant BLAKE3 implementation.

I also felt that there might be some dissatisfaction with the approach given how the bindings were defined since there wasn't much support for automation (e.g., around bindings for the hash types in either direction). `bindgen` is generally not appropriate for this sort of thing since it often generates incorrect and unsafe code for C++ and even using `cxx` still requires a lot of boilerplate.

To give a bit more context, I've been working on high quality LLVM and MLIR bindings for Rust, using the libtooling and unstable C++ APIs, via a new approach that leverages C++ type traits and metaprogramming facilities, and wanted to try to apply that approach here to allow for more extensibility and less boilerplate. This new approach allows for being able to derive Rust traits for the various Nix C++ types automatically, for example.

But it's been awhile since I've worked on adapting that for the Nix Rust BLAKE3 branch, so maybe it would be better in the meantime to revive the earlier Rust version I did have working for Nix, and create a draft PR based on that, at least to serve as a reference point for discussion about some of the implementation issues that arise.

Then maybe later on I can elaborate more on the ideas I had to make things nicer for broader interoperability between C++ and Rust in the code base.

The code I had is based on the Nix repo from some months ago, before the Meson changes were starting to be introduced, so I'm not sure what it will be like to bring that back up to date. But I'll try to open a draft PR if I can get things building again.

**Comment by Ericson2314** [comment]
@silvanshade I sort of feel it might be best just to land the C thing after all?

I have nothing against Rust and better bindgen --- I like those things a lot! --- but I would rather land some simple-stupid-slow BLAKE3 now, and replace it with something better later :). This is especially as since we just switched to Meson, we are perhaps a bit "fatigued" on process workflow changes in the short term (emphasis on short term!), if that makes any sense.

I don't think there will be a perfect-vs-good situation here --- if Blake hashing becomes popular, then I would expect more *desire* for better perf, more fun things from Tvix, etc., not any complacency that the existing C stuff is "good enough".

(Separately, I would encourage you to take your bindgetn expertise to the Rust bindings in https://github.com/nixops4/nixops4. Completely independently of BLAKE, if we can do Rust things better, we can stop C++-ing sooner :))

**Comment by silvanshade** [comment]
@Ericson2314 Okay, that sounds fine to me.

I can re-open and update the linked PR for the C implementation.

**Comment by Ericson2314** [comment]
@silvanshade that would be fantastic! Please do! :)

**Comment by Ericson2314** [comment]
OK this has landed! Per https://github.com/NixOS/nix/pull/12379#issuecomment-2637977897, nothing technical is needed more for stabilization. Rather, we are just waiting for see how much traction this gets in the wild.

**Comment by silvanshade** [comment]
It looks like there is no stabilization issue for BLAKE3.

Given the issues described in https://github.com/NixOS/nix/issues/13511 is there any reason to not just allow the feature as stable?

**Comment by mschwaig** [comment]
The experimental feature seems to be about SRI hashes, as used in `fetchUrl` for example, which according to https://github.com/NixOS/nix/pull/12379#issuecomment-2623186988 Snix does not support.

Where Snix uses BLAKE3 is in its storage layer `castore`.

So I think if the goal is compatibility with Snix then what's more relevant is work on compatibility with the storage layer.
In that context, I want to mention that my proposal for a signature format, https://github.com/mschwaig/laut, opts to put multiple hashes into the output section of a signature to facilitate interoperability and migrations like that in the ecosystem.

We already put both a hash of Snix's `castore-entry` and Nix's current format content hash of a path into the same signature, powered by Snix, but done in a the `post-build-hook` for content-addressed derivations. I would like to upstreamed in the future, but lots of stuff still has to happen first. The most pressing is support for input-addressed derivations, but with inputs and outputs hashed as if the derivations were content addressed.

This is what that feature looks like in the decoded signature of a `hello` derivation: 
```
{
    "in": {
        "rdrv_aterm_ca": "7gsw68f2iawn9q8vv04zh9xql14268pw",
        "debug": {
            "drv_name": "hello-2.12.1",
            "rdrv_path": "/nix/store/7gsw68f2iawn9q8vv04zh9xql14268pw-hello-2.12.1.drv",
            "rdrv_json_preimage": [...],
            "rdrv_computed_path": "/nix/store/7gsw68f2iawn9q8vv04zh9xql14268pw-hello-2.12.1.drv",
            "rdrv_aterm_ca_preimage": [...]
        }
    },
    "out": {
        "castore-entry": {
            "out": "CiUSIDdTvGaPgk_V75yYiefkznYySNlB6yHj_nRKaEONGOLcGJoC"
        },
        "nix": {
            "out": {
                "hashAlgo": "r:sha256",
                "path": "/nix/store/56zjcp964v7lcigv4fjgw0aqgyqpz75h-hello-2.12.1",
                "hash": "sha256:18ipw7vwcixhwp8g1rma6lgl28zcj6mqyj1vx50pj7pkp81kr6c5"
            }
        }
    },
    "builder": {
        "rebuild_id": 2490310005,
        "store_root": "/nix/store",
        "nix_flavor": "lix",
        "nix_version": "2.91.1"
    }
}
```

On the same level as `rdrv_aterm_ca`, which is the hash of a resolved content-addressed derivation, something like the content-hash of a `BuildRequest` from Snix could also identify the set of inputs.

**Comment by silvanshade** [comment]
> The experimental feature seems to be about SRI hashes, as used in `fetchUrl` for example, which according to [#12379 (comment)](https://github.com/NixOS/nix/pull/12379#issuecomment-2623186988) Snix does not support.

Yes, that was the feature landed in the PR mentioned in @Ericson2314's earlier post: https://github.com/NixOS/nix/issues/11999#issuecomment-2638761977

> Where Snix uses BLAKE3 is in its storage layer `castore`.
> 
> So I think if the goal is compatibility with Snix then what's more relevant is work on compatibility with the storage layer.

This idea is interesting and worth pursuing but it's probably better to create a new issue about it to clearly outline what the specific goals are and a roadmap for how to achieve them.

Although the original post in this thread mentions tvix as one of the motivations for BLAKE3 support, all of the subsequent work that has been done, and the PRs merged, are only about making it usable as another general purpose hash algorithm.

> In that context, I want to mention that my proposal for a signature format, https://github.com/mschwaig/laut, opts to put multiple hashes into the output section of a signature to facilitate interoperability and migrations like that in the ecosystem.
> 
> We already put both a hash of Snix's `castore-entry` and Nix's current format content hash of a path into the same signature, powered by Snix, but done in a the `post-build-hook` for content-addressed derivations. I would like to upstreamed in the future, but lots of stuff still has to happen first. The most pressing is support for input-addressed derivations, but with inputs and outputs hashed as if the derivations were content addressed.

These sound like some good ideas, especially the option for multiple hash digests, but I think that's an entirely separate development and shouldn't impact whether to allow BLAKE3 SRI hashes in the stable interface.

**Comment by flokli** [comment]
> These sound like some good ideas, especially the option for multiple hash digests, but I think that's an entirely separate development and shouldn't impact whether to allow BLAKE3 SRI hashes in the stable interface.

BLAKE3 is not part of the [SRI spec](https://w3c.github.io/webappsec-subresource-integrity/), and neither is md5 btw.

We could guess the hash prefix is "b3-", or "blake3-" (not sure what was picked in the end) but are we comfortable stabilizing it without knowing how it'll end up in the SRI spec?

**Comment by silvanshade** [comment]
> BLAKE3 is not part of the [SRI spec](https://w3c.github.io/webappsec-subresource-integrity/), and neither is md5 btw.

That's a fair point. But since Nix already allows other algorithms in SRIs, I wonder if it really matters much.

There's also the possibility that BLAKE3 is never included in the official SRI spec. If we wait on that, people may never get to use the feature except in special cases.

The [BLAKE3 draft](https://www.ietf.org/archive/id/draft-aumasson-blake3-00.html) was not completed, so I would actually be surprised to see the algorithm included in the official spec.

> not sure what was picked in the end

Here it is just `blake3-`.

> are we comfortable stabilizing it without knowing how it'll end up in the SRI spec?

What is really the worst case scenario here though?

If we were to stabilize it and somehow something relevant changes in the SRI spec that requires revisiting some part of the implementation or how its referenced in some packages, isn't that something we can relatively easily fix?

**Comment by Lord-Valen** [comment]
> OK this has landed! Per [#12379 (comment)](https://github.com/NixOS/nix/pull/12379#issuecomment-2637977897), nothing technical is needed more for stabilization. Rather, we are just waiting for see how much traction this gets in the wild.

What exactly constitutes "actual use"? How much use and by whom? I know [nixtamal](https://nixtamal.toast.al/) counts BLAKE3 support among its key features.
